storage unit test flakes with "Unexpected error: too old resource version: 2 (3)"

Looks like a unit test flake. A race between two concurrent instances of this test updating the same resource?
```
<testsuite tests="9" failures="1" time="1.451" name="k8s.io/kubernetes/pkg/storage">
<properties>
<property name="go.version" value="go1.4.3"/>
</properties>
<testcase classname="storage" name="TestEtcdParseWatchResourceVersion" time="0.000"/>
<testcase classname="storage" name="TestWatchCacheBasic" time="0.000"/>
<testcase classname="storage" name="TestEvents" time="0.000"/>
<testcase classname="storage" name="TestWaitUntilFreshAndList" time="0.000"/>
<testcase classname="storage" name="TestReflectorForWatchCache" time="0.000"/>
<testcase classname="storage" name="TestList" time="0.270"/>
<testcase classname="storage" name="TestWatch" time="0.370"/>
<testcase classname="storage" name="TestWatcherTimeout" time="0.160">
<failure message="Failed" type="">
cacher_test.go:235: Unexpected error: too old resource version: 2 (3)
</failure>
</testcase>
<testcase classname="storage" name="TestFiltering" time="0.380"/>
</testsuite>
```

cc @saad-ali

OK, I've got something:

    === RUN   TestWatch
    --- FAIL: TestWatch (0.08s)
	    cacher_test.go:184: Unexpected error: too old resource version: 1 (3)

https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/pr-logs/9d912e843b40193c59cfa8d72b6b5780bda53b11/kubernetes-pull-test-unit-integration/7414/?debugUI=DEVELOPERS

--- a/pkg/storage/cacher.go
+++ b/pkg/storage/cacher.go
@@ -293,7 +293,7 @@ func (c *Cacher) List(ctx context.Context, key string, resourceVersion string, f
 		return err
 	}
 
-	// To avoid situation when List is proceesed before the underlying
+	// To avoid situation when List is processed before the underlying
 	// watchCache is propagated for the first time, we acquire and immediately
 	// release the 'usable' lock.
 	// We don't need to hold it all the time, because watchCache is thread-safe
@@ -399,6 +399,11 @@ func filterFunction(key string, keyFunc func(runtime.Object) (string, error), fi
 
 // Returns resource version to which the underlying cache is synced.
 func (c *Cacher) LastSyncResourceVersion() (uint64, error) {
+	// To avoid situation when LastSyncResourceVersion is processed before the
+	// underlying watchCache is propagated, we acquire 'usable' lock.
+	c.usable.RLock()
+	defer c.usable.RUnlock()
+
 	c.RLock()
 	defer c.RUnlock()
 

--- a/pkg/storage/cacher_test.go
+++ b/pkg/storage/cacher_test.go
@@ -178,8 +178,16 @@ func TestWatch(t *testing.T) {
 	podFooBis := makeTestPod("foo")
 	podFooBis.Spec.NodeName = "anotherFakeNode"
 
+	// initialVersion is used to initate the watcher at the beginning of the world,
+	// which is not defined precisely in etcd.
+	initialVersion, err := cacher.LastSyncResourceVersion()
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+	startVersion := strconv.Itoa(int(initialVersion))
+
 	// Set up Watch for object "podFoo".
-	watcher, err := cacher.Watch(context.TODO(), "pods/ns/foo", "1", storage.Everything)
+	watcher, err := cacher.Watch(context.TODO(), "pods/ns/foo", startVersion, storage.Everything)
 	if err != nil {
 		t.Fatalf("Unexpected error: %v", err)
 	}
@@ -227,15 +235,23 @@ func TestWatcherTimeout(t *testing.T) {
 	cacher := newTestCacher(etcdStorage)
 	defer cacher.Stop()
 
+	// initialVersion is used to initate the watcher at the beginning of the world,
+	// which is not defined precisely in etcd.
+	initialVersion, err := cacher.LastSyncResourceVersion()
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+	startVersion := strconv.Itoa(int(initialVersion))
+
 	// Create a watcher that will not be reading any result.
-	watcher, err := cacher.WatchList(context.TODO(), "pods/ns", "1", storage.Everything)
+	watcher, err := cacher.WatchList(context.TODO(), "pods/ns", startVersion, storage.Everything)
 	if err != nil {
 		t.Fatalf("Unexpected error: %v", err)
 	}
 	defer watcher.Stop()
 
 	// Create a second watcher that will be reading result.
-	readingWatcher, err := cacher.WatchList(context.TODO(), "pods/ns", "1", storage.Everything)
+	readingWatcher, err := cacher.WatchList(context.TODO(), "pods/ns", startVersion, storage.Everything)
 	if err != nil {
 		t.Fatalf("Unexpected error: %v", err)
 	}

